### PR TITLE
Simplify node configuration

### PR DESCRIPTION
**Pull request by neilalexander**
This PR really grew arms and legs from where I first started. It basically:

- Completely removes all `NodeConfig` references from `yggdrasil`, `multicast`, `tuntap`, `admin`, etc
- Adds and improves some mutator functions on these packages (all configuration changes will be done through these methods now)
- Various things have been shifted around
- Some new actors have been created (e.g. `peers`)

Still to do:
- [ ] Soft reconfiguration support in `cmd/yggdrasil`
- [ ] Session firewall might not be safe
- [ ] Test to see if everything still works
- [ ] Improve persistence of API functions for adding/removing peers etc

### CODE REVIEW

File: src/yggdrasil/api.go [hunk 1]
```diff
@@ -370,13 +370,12 @@ func (c *Core) AddPeer(addr string, sintf string) error {
 	if err := c.CallPeer(addr, sintf); err != nil {
 		return err
 	}
-	c.config.Mutex.Lock()
+	config := c.GetConfig()
 	if sintf == "" {
-		c.config.Current.Peers = append(c.config.Current.Peers, addr)
+		config.Peers = append(config.Peers, addr)
```
Comment (Arceliar) [inline src/yggdrasil/api.go]: This isn't thread safe, since it mutates the existing config.

File: src/yggdrasil/api.go [hunk 2]
```diff
@@ -370,13 +370,12 @@ func (c *Core) AddPeer(addr string, sintf string) error {
...
 	} else {
-		c.config.Current.InterfacePeers[sintf] = append(c.config.Current.InterfacePeers[sintf], addr)
+		config.InterfacePeers[sintf] = append(config.InterfacePeers[sintf], addr)
```
Comment (Arceliar) [inline src/yggdrasil/api.go]: This isn't thread safe, since it mutates the existing config.

File: src/yggdrasil/peer.go [hunk 3]
```diff
@@ -34,50 +35,42 @@ func (ps *peers) init(c *Core) {
...
 }
 
 // Adds a key to the whitelist.
 func (ps *peers) addAllowedEncryptionPublicKey(box string) {
-	ps.core.config.Mutex.RLock()
-	defer ps.core.config.Mutex.RUnlock()
-	ps.core.config.Current.AllowedEncryptionPublicKeys =
-		append(ps.core.config.Current.AllowedEncryptionPublicKeys, box)
+	current := ps.core.GetConfig()
+	current.AllowedEncryptionPublicKeys = append(current.AllowedEncryptionPublicKeys, box)
```
Comment (Arceliar) [inline src/yggdrasil/peer.go]: This isn't thread safe, since it mutates the existing config.

File: src/yggdrasil/peer.go [hunk 4]
```diff
@@ -34,50 +35,42 @@ func (ps *peers) init(c *Core) {
...
-	ps.core.config.Mutex.RLock()
-	defer ps.core.config.Mutex.RUnlock()
-	for k, v := range ps.core.config.Current.AllowedEncryptionPublicKeys {
+	current := ps.core.GetConfig()
+	for k, v := range current.AllowedEncryptionPublicKeys {
 		if v == box {
-			ps.core.config.Current.AllowedEncryptionPublicKeys =
-				append(ps.core.config.Current.AllowedEncryptionPublicKeys[:k],
-					ps.core.config.Current.AllowedEncryptionPublicKeys[k+1:]...)
+			current.AllowedEncryptionPublicKeys = append(current.AllowedEncryptionPublicKeys[:k], current.AllowedEncryptionPublicKeys[k+1:]...)
```
Comment (Arceliar) [inline src/yggdrasil/peer.go]: This isn't thread safe, since it mutates the existing config.

**Comment by neilalexander** [comment]
I'm going to close this for now because I think it's obvious that we need to finish actor-ising everything before a lot of this can be unpicked. This was supposed to be a small set of changes which ended up having sprawling and unforeseen effects and I'm still not convinced it's all fixable right now.

**Pull request closed without merging**